virtool/uploads/api.py

```diff
@@ -53,59 +58,136 @@ async def naive_writer(req, file_id):
             size += len(chunk)
             await handle.write(chunk)
 
-    return size
+    return size, virtool.utils.timestamp()
 
 
-@routes.post("/upload/{file_type}", permission="upload_file")
+@routes.post("/api/uploads", permission="upload_file")
 async def upload(req):
-    db = req.app["db"]
-
-    file_type = req.match_info["file_type"]
-
-    if file_type not in FILE_TYPES:
-        return not_found()
+    db = req.app["postgres"]
+    upload_type = req.query.get("type")
 
     errors = naive_validator(req)
 
     if errors:
         return invalid_query(errors)
 
-    filename = req.query["name"]
+    name = req.query["name"]
 
-    document = await virtool.files.db.create(
-        db,
-        filename,
-        file_type,
-        user_id=req["client"].user_id
-    )
+    if upload_type not in UPLOAD_TYPES:
+        return bad_request("Unsupported upload type")
 
-    file_id = document["id"]
+    document = await virtool.uploads.db.create(db, name, upload_type, user=req["client"].user_id)
 
-    try:
-        size = await naive_writer(req, file_id)
+    upload_id = document["id"]
+
+    async with AsyncSession(db) as session:
+        try:
+            result = (await session.execute(select(Upload).filter_by(id=upload_id))).scalar()
+            size, uploaded_at = await naive_writer(req, upload_id, name)
+
+            result.size = size
+            result.name_on_disk = f"{upload_id}-{name}"
+            result.uploaded_at = uploaded_at
+
+            await session.commit()
 
-        await db.files.update_one({"_id": file_id}, {
-            "$set": {
+            document.update({
                 "size": size,
-                "ready": True
+                "name_on_disk": f"{upload_id}-{name}",
+                "uploaded_at": uploaded_at
+            })
+
+            logger.debug(f"Upload succeeded: {upload_id}")
+
+            headers = {
+                "Location": f"/api/uploads/{upload_id}"
             }
-        })
```

Comment:
Please get rid of references to `document` here. This is a MongoDB concept.

You should return the model object from `db.create()` then use `to_dict()` to get the dictionary representation of the data.